```javascript
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
<!--<a href="javaScript:alert('清除成功!')">清除缓存</a>-->
<script>
    /*
        异常:
            程序执行不止有正常执行结果  还会有很多执行出现不正常的情况
            这些不正常的情况统一看做异常

            我们要针对程序的异常要做处理 否则 程序很不健全

            前端一般常见的异常情况:
                    运行环境不同带来的异常(浏览器兼容);
                    代码因为计算或者其他逻辑方面问题带来的错误
                    语法错误


       举个栗子:
            连宝同学骑行去爬山

                1.骑到一半  发现 车胎没气了    (出发前就应该检查好的)

                2.骑到路途中,发现路边有好多小石子  很好玩  骑行压石子
                        嘭的一声  车翻了   (脑残问题)


                3.骑到半山腰   遇到山体滑坡  只能原路返回(不能做针对性处理)

           写代码期间的编译期错误(做预处理  类似于防火)

           代码运行期间出现的错误做处理(如果不处理 那么久报错终止)

           处理问题两种方式:
                第一种:  抛出异常  直接终止程序
                    throw new Error("错误提示")

                第二种:  try{}catch(){}代码块  做针对性处理
                    针对有可能出现的错误做预处理

                    格式:
                           try{
                                尝试执行的代码

                           }catch(err){//捕获到的异常对象
                            针对出现的异常做处理

                           }

                    顺序代码

                执行顺序:
                    a.先尝试执行try代码块里面的代码
                    b.如果没有报错 那么 顺序向下执行其他代码 不执行catch
                    c.如果try里面出现报错 系统自动捕获异常信息对象 执行catch代码块
                    d.catch代码块 程序员手动进行异常处理
                    e.catch执行完 顺序向下执行其他代码

                    优点: 一旦有try catch代码块 那么  程序如果出现报错
                    后面代码继续执行

                try catch无法捕获语法错误

                throw  还可以直接跟有一个json对象


          异常是有传递机制的:
                如果调用的方法里面出现了异常 那么会一直打印错误信息到最上层调用者


          finally语句:
                无论如何都会执行的代码块

          try{

           }catch(e){

           }finally{
                无论如何都会执行的代码块  一般情况下可以省略不写

           }

     */

    /*var boo=true;
    if(boo){
        throw new Error("你网页崩溃了...");

    }*/

       /* try{

            var str=null;

            console.log(str.length);

        }catch(e){
//            console.error(e);
            //这种只是打印了错误信息 并没有针对性处理  那么该如何处理问题
            //要具体问题具体分析
            str="处理后的值";
            console.log(str.length);
            
        }

    alert("我很重要,死也要执行的代码!");*/



    /*try{
//        语法错误无法捕获
        var str="hsfdgh;

    }catch(e){

    }*/

//    抛出一个自定义对象  程序依然会报错终止
    /*throw {code:4444,msg:"错误鸭!!"};

    alert("1")*/
/*
    function f3() {
        f2();
        
    }
    
    

    function f2() {
        f1();

    }

    
    function f1() {
        throw {code:4444,msg:"错误鸭!!"};
        
    }

    f3();*/


    function fn() {
        try{
            console.log("程序开始");
            var str="xx";

            console.log(str.length);
            console.log("程序没有报错执行我!");
            return "程序正确";

        }catch(e){
            console.error("用输出语句代替处理错误!");

//            return "程序错误"
        }finally{
            //如果try和catch代码块中没有 return语句 那么finally可以省略不写
            //相反 如果有  那么finally必须写
            console.log("我是无论如何都必须执行的代码");
        }
        

        console.log("我是顺序结构的代码,我也应该执行");
    }
    fn();
</script>
</body>
</html>
```

